arch/arm/core/aarch64/userspace.S

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (c) 2020 Carlo Caione <ccaione@baylibre.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <toolchain.h>
+#include <linker/sections.h>
+#include <offsets_short.h>
+#include <arch/cpu.h>
+#include <syscall.h>
+#include "macro_priv.inc"
+
+_ASM_FILE_PROLOGUE
+
+/*
+ * size_t arch_user_string_nlen(const char *s, size_t maxsize, int *err_arg)
+ */
+
+GTEXT(z_arm64_user_string_nlen_fault_start)
+GTEXT(z_arm64_user_string_nlen_fault_end)
+GTEXT(z_arm64_user_string_nlen_fixup)
+
+GTEXT(arch_user_string_nlen)
+SECTION_FUNC(TEXT, arch_user_string_nlen)
+
+	mov	x3, x0
+	mov	x0, #0
+	mov	x4, #0
+
+strlen_loop:
+
+	cmp	x0, x1
+	beq	strlen_done
+
+z_arm64_user_string_nlen_fault_start:
+	ldrb	w5, [x3, x0]
+z_arm64_user_string_nlen_fault_end:
+	cbz	x5, strlen_done
+
+	add	x0, x0, #1
+	b	strlen_loop
+
+z_arm64_user_string_nlen_fixup:
+	mov	x4, #-1
+	mov	x0, #0
+
+strlen_done:
+	str	w4, [x2]
+	ret
+
+/*
+ * int arch_buffer_validate(void *addr, size_t size, int write)
+ */
+
+GTEXT(arch_buffer_validate)
+SECTION_FUNC(TEXT, arch_buffer_validate)
+
+	add	x1, x1, x0
```

Comment:
This loop doesn't appear to check non-page-aligned addresses/sizes properly.

E.g. if I am checking 4KiB starting halfway through a page it only checks the first of the two pages, assuming I'm parsing this correctly.

This may be worth tossing in a test for.

Suggestion:

1. At the start of the loop, do `and x0, x0, ~#(CONFIG_MMU_PAGE_SIZE - 1)` to reset the address to the start of the page.
2. Change the increment to `add x0, x0, #(CONFIG_MMU_PAGE_SIZE)`.

(This won't work if x0 is 2**64-CONFIG_MMU_PAGE_SIZE, but that can't happen in practice.)